File: api/gnrs_api_example.php
<?php

// ************************************************************
// WARNING: Not up to date! See gnrs_api_example.R for new calls
// ************************************************************

//////////////////////////////////////////////////////
// Submits test data to GNRS API & displays results
//
// * Imports names from test file on local file system
// * Displays input and output at various stages of
// 	 the process.
// * 
//////////////////////////////////////////////////////

/////////////////////////////////////////
// API processing options
/////////////////////////////////////////

// Processing mode
$mode="resolve";		// Resolve names
$mode="countrylist";	// Return information on all countries
$mode="statelist";		// Return info on all states; country_id param required
$mode="countylist";		// Return info on all counties; state_id param required
$mode="meta";			// Return metadata on application & sources

// Number of batches for parallel processing
$ppbatches=20;

/////////////////////
// API parameters
/////////////////////

// api base url 
$base_url = "http://vegbiendev.nceas.ucsb.edu:8875/gnrs_api.php"; // production
$base_url = "http://vegbiendev.nceas.ucsb.edu:9875/gnrs_api.php"; // development
//$base_url = "http://paramo.cyverse.org/gnrs/gnrs_api.php";        // paramo

require_once 'params.php';			// general parameters 
require_once 'api_params.php';		// api-specific parameters

// Path and name of file containing input names and political divisions
$inputfile = $DATADIR."gnrs_testfile.csv";	// local test file
//$inputfile = "https://bien.nceas.ucsb.edu/bien/wp-content/uploads/2020/11/gnrs_testfile.csv";

// Desired response format
//	Options: json*|xml
// Example: $format="xml";
// NOT YET IMPLEMENTED!
$format="json";

// Number of lines to import
// Use this option to limit test data to small subsample of input file
// Set to number > # of lines in file to import entire file
$lines = 10000000000;
//$lines = 5;

//////////////////////////////////////////
// Test data from routes statelist & countylist
//////////////////////////////////////////

// List of country_ids 'Costa Rica, Nicaragua, Panama)
$country_ids="3624060,3617476,3703430";
$country_ids="";

// List of example state_province IDs
$state_ids="3624953,3624368,3830308,3620673";
//$state_ids="";
$state_ids="3624953,3624368,3830308,hello world!";

/////////////////////////////////////////
// Display options
// 
// * Turn on/off what is echoed to terminal
// * Raw data always displayed
/////////////////////////////////////////

$disp_data_array=false;		// Echo raw data as array
$disp_combined_array=false;	// Echo combined options+data array
$disp_opts_array=false;		// Echo application options as array
$disp_opts=true;			// Echo application options
$disp_json_data=true;		// Echo the options + raw data JSON POST data
$disp_results_json=true;	// Echo results as array
$disp_results_array=false;	// Echo results as array
$disp_results_csv=true;		// Echo results as CSV text
$time=true;					// Echo time elapsed

/////////////////////////////////////////////////////
// Command line options
// Use to over-ride the above parameters
/////////////////////////////////////////////////////

// Get options, set defaults for optional parameters
// Use default if unset
$options = getopt("m:b:");
$mode=isset($options["m"])?$options["m"]:$mode;	
$batches=isset($options["b"])?$options["b"]:$ppbatches;	

////////////////////////////////////////////////////////////////
// Main
////////////////////////////////////////////////////////////////

include $timer_on; 	// Start the timer
echo "\n";

///////////////////////////////
// Set options array
///////////////////////////////
$opts_arr = array(
	"mode"=>$mode
	);
if ( $mode=="resolve" && (! $batches=="" ) ) {
	$opts_arr += array("batches"=>$batches);
}
	
///////////////////////////////
// Display options and confirm
///////////////////////////////

if ($disp_opts) {	
	// Echo the options
	echo "$APPNAME options:\r\n";
	foreach($opts_arr as $key => $value) {
		//foreach($row as $key => $value) {
			echo "  $key=$value\n";
		//}
	}
	echo $mode=="resolve"?"  input file: $inputfile \n":"";
	echo "  API url: $base_url \n";
	echo "\r\n";
}

$batches=isset($options["b"])?$options["b"]:$ppbatches;	


// Confirm above options before proceeding
$message   =  "Execute API call with above settings? [y/n]";
print $message;
// flush();
// ob_flush();
$confirmation  =  trim( fgets( STDIN ) );
if ( $confirmation !== 'y' ) {
   exit (0);
}

///////////////////////////////
// Make data array
///////////////////////////////

if ( $mode=="resolve" ) {
	// Import csv data and convert to array
	$data_arr = array_map('str_getcsv', file($inputfile));

	# Get subset
	$data_arr = array_slice($data_arr, 0, $lines);

	$lines_in = 0;

	// Echo raw data
	echo "The raw data:\r\n";
	foreach($data_arr as $row) {
		foreach($row as $key => $value) echo "$value\t"; echo "\r\n";
		$lines_in++;
	}
	echo "\r\n";
	
	$lines_in = $lines_in-1;
	echo "\nInput rows (minus header): $lines_in\n\n";

	if ($disp_data_array) {
		echo "The raw data as array:\r\n";
		var_dump($data_arr);
		echo "\r\n";
	}
} else if ( $mode=="statelist" ) {
	$val_arr=explode(",",$country_ids);
	foreach($val_arr as $value) {
		$data_arr[] = array('country_id' => $value);
	}	
} else if ( $mode=="countylist" ) {
	$val_arr=explode(",",$state_ids);
	foreach($val_arr as $value) {
		$data_arr[] = array('state_id' => $value);
	}	
}

///////////////////////////////
// Merge options and data into 
// json object for post
///////////////////////////////

// Convert to JSON
if ($mode=='resolve' || $mode=='statelist' || $mode=='countylist' ) {
	# Options + data
	$json_data = json_encode(array('opts' => $opts_arr, 'data' => $data_arr));	
} else {
	# Just options
	$json_data = json_encode(array('opts' => $opts_arr));	
}

///////////////////////////////
// Decompose the JSON
// into opt and data
///////////////////////////////

$input_array = json_decode($json_data, true);

if ($disp_combined_array) {
	echo "The combined array:\r\n";
	var_dump($input_array);
	echo "\r\n";
}

$opts = $input_array['opts'];
if ($disp_opts_array) {
	echo "Options array:\r\n";
	var_dump($opts);
	echo "\r\n";
}

if ($disp_json_data) {
	// Echo the final JSON post data
	echo "API input (options + raw data converted to JSON):\r\n";
	echo $json_data . "\r\n\r\n";
}

///////////////////////////////
// Call the API
///////////////////////////////

$url = $base_url;    

// Initialize curl & set options
$ch = curl_init($url);	
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);	// Return response (CRITICAL!)
curl_setopt($ch, CURLOPT_POST, 1);	// POST request
curl_setopt($ch, CURLOPT_POSTFIELDS, $json_data);	// Attach the encoded JSON
curl_setopt($ch, CURLOPT_HTTPHEADER, array('Content-Type: application/json')); 

// Send the API call
$response = curl_exec($ch);

// Check status of the response and echo if error
$status = curl_getinfo($ch, CURLINFO_HTTP_CODE);

if ( $status != 201 && $status != 200 ) {
	$status_msg = $http_status_codes[$status];
	$msg="Error: call to URL $url failed with status $status $status_msg \r\nDetails: $response \r\n";
    //die($msg);
    echo $msg;
}

// Close curl
curl_close($ch);

///////////////////////////////
// Echo the results
///////////////////////////////

$results_json = $response;
$results = json_decode($results_json, true);	// Convert JSON results to array

// Echo the JSON response
if ($disp_results_json) {
	echo "API results (JSON)\r\n";
	echo $results_json;
	echo "\r\n\r\n";
}

if ($disp_results_array) {
	echo "API results as array:\r\n";
	var_dump($results);
	echo "\r\n\r\n";
}

if ($disp_results_csv) {
	$lines_out = 0;
	echo "API results as CSV:\r\n";
	
	foreach ( $results as $rkey => $row ) {
		$lines_out++;
		$rind=array_search( $rkey, array_keys($results) );	# Index: current row
		$cindmax = count( $row )-1;	// Index: last column of current row
	
		if ( $rind==0 ) {
			// Print header
			foreach ( $row as $key => $value )  {	
				$cind=array_search( $key, array_keys($row) );
				$cind==$cindmax?$format="%1s\n":$format="%1s,";
				printf($format, $key);
			}
		}

		// Print data 
		foreach ( $row as $key => $value )  {	
			$cind=array_search( $key, array_keys($row) );
			$cind==$cindmax?$format="%1s\n":$format="%1s,";
			printf($format, $value);
		}
	}
	
	$lines_out = $lines_out-1;
	echo "\nOutput rows (minus header): $lines_out\n\n";
}

///////////////////////////////////
// Echo time elapsed
///////////////////////////////////

include $timer_off;	// Stop the timer
if ($time) echo "\r\nTime elapsed: " . $tsecs . " seconds.\r\n\r\n"; 

?>

File: api/gnrs_api_example.R
###############################################
# GNRS API Examples
#
# Invokes parallel processing GNRS interface (gnrspar.sh)
###############################################

rm(list=ls())

#################################
# Parameters
#################################

# api url
url = "https://gnrsapi.xyz/gnrs_api.php" 

# Test files of political divisions to resolve (choose one)
# Format: comma-delimited UTF-8 text
# Columns:
#   user_id: optional unique identifier. Can be blank but comma placeholder required
#   country: required
#   state: optional, country required if included
#   county: optional, country and state required if included

# Test data with unique ids
testfile.ids <- "https://bien.nceas.ucsb.edu/bien/wp-content/uploads/2020/11/gnrs_testfile_ids.csv"

# Test data, no ids
testfile.noids <- "https://bien.nceas.ucsb.edu/bien/wp-content/uploads/2020/11/gnrs_testfile.csv"

# Set the input file
testfile <- testfile.ids

# Load libraries
library(httr)		# API requests
library(jsonlite) # JSON coding/decoding

#################################
# Import & prepare the raw data
#################################

# Read in example file of taxon names
data <- read.csv(testfile, header=TRUE)

# Inspect the input data
head(data,25)

# Uncomment to work with smaller sample of the data
#data <- head(data,8)

# Convert the data to JSON
data_json <- jsonlite::toJSON(unname(data))

#################################
# Example 1: Resolve
#################################

# Set API options
mode <- "resolve"			# Processing mode
batches <- 10					# Number of batches, for parallel processing
										# input file will be divided into this many batches

# Convert the options to data frame and then JSON
opts <- data.frame( c(mode) )
names(opts) <- c("mode")
if ( exists("batches") ) opts$batches <- batches

opts_json <-  jsonlite::toJSON(opts)
opts_json <- gsub('\\[','',opts_json)
opts_json <- gsub('\\]','',opts_json)

# Combine the options and data into single JSON object
input_json <- paste0('{"opts":', opts_json, ',"data":', data_json, '}' )

# Send the API request
results_json <- POST(url = url,
                  add_headers('Content-Type' = 'application/json'),
                  add_headers('Accept' = 'application/json'),
                  add_headers('charset' = 'UTF-8'),
                  body = input_json,
                  encode = "json"
                  )

# Convert JSON results to a data frame
results_raw <- fromJSON(rawToChar(results_json$content)) 
results <- as.data.frame(results_raw)

# Inspect the results
head(results, 10)

# Display header plus one row vertically
# to better compare the output fields
results.t <- as.data.frame( t( results[,1:ncol(results)] ) )
results.t[,3,drop =FALSE]

# Display columns showing how each political division level was matched
results[ , c( 'country_verbatim', 'state_province_verbatim', 'county_parish_verbatim', 
	'country', 'state_province', 'county_parish', 
	'match_method_country', 'match_method_state_province', 'match_method_county_parish')
	]

# Display columns showing overall completeness of match
results[ , c('country_verbatim', 'state_province_verbatim', 'county_parish_verbatim', 
	'country', 'state_province', 'county_parish', 'match_method_country', 
	'poldiv_submitted', 'poldiv_matched', 'match_status')
	]

# Display alternative IDs returned in addition to names.
# IDs beginning "gid_" are gadm identifiers
# geonameid refers is for the lowest political division matched
results[ , c(	'country', 'state_province', 'county_parish', 
	'country_iso', 'state_province_iso', 'county_parish_iso', 
	'gid_0', 'gid_1', 'gid_2', 'geonameid' )
	]

#################################
# Example 2: Get list of all countries & 
# associated information in GNRS DB
# Parameters: none
#################################
rm( list = Filter( exists, c("results", "results_json") ) )

# All we need to do is reset option mode.
# all other options will be ignored
mode <- "countrylist"		

# Re-form the options json again
# Note that only 'mode' is needed
opts <- data.frame(c(mode))
names(opts) <- c("mode")
opts_json <- jsonlite::toJSON(opts)
opts_json <- gsub('\\[','',opts_json)
opts_json <- gsub('\\]','',opts_json)

# Input json requires only the option
input_json <- paste0('{"opts":', opts_json, '}' )

# Send the API request
results_json <- POST(url = url,
                  add_headers('Content-Type' = 'application/json'),
                  add_headers('Accept' = 'application/json'),
                  add_headers('charset' = 'UTF-8'),
                  body = input_json,
                  encode = "json"
                  )

# Convert JSON results to a data frame
results <- as.data.frame( fromJSON( rawToChar( results_json$content ) ) )

# Display the results
print( results )

# Save list of countries for use in state query
countries.all <- results
	
#################################
# Example 3: Get list of states & associated
# information for subset of countries
# Parameters: one or more value of country_id
#################################
rm( list = Filter( exists, c("results", "results_json") ) )

# Make list of countries for which you'd like to get list of states
# Filter however you want, but value used for data_json
# must be the integer country_id
countries <- as.data.frame(countries.all[ countries.all$country 
	%in% c('Costa Rica', 'Nicaragua', 'Panama'), c('country_id')])
# Or set to empty string ("") to get everything:
#countries <- as.data.frame("")	
names(countries ) <- c("country_id")
data_json <- jsonlite::toJSON(countries)

mode <- "statelist"		
opts <- data.frame(c(mode))
names(opts) <- c("mode")
opts_json <- jsonlite::toJSON(opts)
opts_json <- gsub('\\[','',opts_json)
opts_json <- gsub('\\]','',opts_json)

# Form the input json, including both options and data
input_json <- paste0('{"opts":', opts_json, ',"data":', data_json, '}' )

# Send the API request
results_json <- POST(url = url,
                  add_headers('Content-Type' = 'application/json'),
                  add_headers('Accept' = 'application/json'),
                  add_headers('charset' = 'UTF-8'),
                  body = input_json,
                  encode = "json"
                  )

# Convert JSON results to a data frame
results <- as.data.frame( fromJSON( rawToChar( results_json$content ) ) )

# Display the results
print( results )		
print( paste0( "Total rows returned: ", nrow(results) ) )

# Save list of states for use in county query
states.all <- results

#################################
# Example 4: Get list of counties & associated
# information for subset of states in one or
# more countries
# Parameters: one or more pairs of 
# country_id + state_province_id
#################################
rm( list = Filter( exists, c("results", "results_json") ) )

# Make list of states for which you'd like to get list of counties
# Filter however you want, but value used for data_json parameters
# must be the integer state_province_id
states <- as.data.frame(states.all[ states.all$state_province_id 
	%in% c(3624953, 3624368, 3830308, 3620673), c('state_province_id')])
# Or set to empty string ("") to get everything:
#states <- as.data.frame("")	
names(states ) <- c("state_province_id")
data_json <- jsonlite::toJSON(states)

mode <- "countylist"		
opts <- data.frame(c(mode))
names(opts) <- c("mode")
opts_json <- jsonlite::toJSON(opts)
opts_json <- gsub('\\[','',opts_json)
opts_json <- gsub('\\]','',opts_json)

# Form the input json, including both options and data
input_json <- paste0('{"opts":', opts_json, ',"data":', data_json, '}' )

# Send the API request
results_json <- POST(url = url,
                  add_headers('Content-Type' = 'application/json'),
                  add_headers('Accept' = 'application/json'),
                  add_headers('charset' = 'UTF-8'),
                  body = input_json,
                  encode = "json"
                  )

# Convert JSON results to a data frame
results <- as.data.frame( fromJSON( rawToChar( results_json$content ) ) )

# Display the results
print( results )
print( paste0( "Total rows returned: ", nrow(results) ) )

#################################
# Example 5: Get data dictionary of GNRS 
# output
#################################
rm( list = Filter( exists, c("results", "results_json") ) )

# All we need to do is reset option mode.
# all other options will be ignored
mode <- "dd"		

# Re-form the options json again
# Note that only 'mode' is needed
opts <- data.frame(c(mode))
names(opts) <- c("mode")
opts_json <- jsonlite::toJSON(opts)
opts_json <- gsub('\\[','',opts_json)
opts_json <- gsub('\\]','',opts_json)

# Make the options
# No data needed
input_json <- paste0('{"opts":', opts_json, '}' )

# Send the API request
results_json <- POST(url = url,
                  add_headers('Content-Type' = 'application/json'),
                  add_headers('Accept' = 'application/json'),
                  add_headers('charset' = 'UTF-8'),
                  body = input_json,
                  encode = "json"
                  )

# Convert JSON results to a data frame
results <- as.data.frame( fromJSON( rawToChar( results_json$content ) ) )

# Display the results
print( results )

#################################
# Example 6: Get metadata for current 
# GNRS version
#################################
rm( list = Filter( exists, c("results", "results_json") ) )

# All we need to do is reset option mode.
# all other options will be ignored
mode <- "meta"		

# Re-form the options json again
# Note that only 'mode' is needed
opts <- data.frame(c(mode))
names(opts) <- c("mode")
opts_json <- jsonlite::toJSON(opts)
opts_json <- gsub('\\[','',opts_json)
opts_json <- gsub('\\]','',opts_json)

# Make the options
# No data needed
input_json <- paste0('{"opts":', opts_json, '}' )

# Send the API request
results_json <- POST(url = url,
                  add_headers('Content-Type' = 'application/json'),
                  add_headers('Accept' = 'application/json'),
                  add_headers('charset' = 'UTF-8'),
                  body = input_json,
                  encode = "json"
                  )

# Convert JSON results to a data frame
results <- as.data.frame( fromJSON( rawToChar( results_json$content ) ) )

# Display the results
print( results )

#################################
# Example 7: Get metadata for all 
# GNRS sources
#################################
rm( list = Filter( exists, c("results", "results_json") ) )

# Set sources mode
mode <- "sources"		

# Re-form the options json again
# Note that only 'mode' is needed
opts <- data.frame(c(mode))
names(opts) <- c("mode")
opts_json <- jsonlite::toJSON(opts)
opts_json <- gsub('\\[','',opts_json)
opts_json <- gsub('\\]','',opts_json)

# Make the options
# No data needed
input_json <- paste0('{"opts":', opts_json, '}' )

# Send the API request
results_json <- POST(url = url,
                  add_headers('Content-Type' = 'application/json'),
                  add_headers('Accept' = 'application/json'),
                  add_headers('charset' = 'UTF-8'),
                  body = input_json,
                  encode = "json"
                  )

# Convert JSON results to a data frame
results <- as.data.frame( fromJSON( rawToChar( results_json$content ) ) )

# Display the results
print( results )

#################################
# Example 8: Get bibtex citations for GNRS 
# data sources and the GNRS
#################################
rm( list = Filter( exists, c("results", "results_json") ) )

# Set citations mode
mode <- "citations"		

# Re-form the options json again
# Note that only 'mode' is needed
opts <- data.frame(c(mode))
names(opts) <- c("mode")
opts_json <- jsonlite::toJSON(opts)
opts_json <- gsub('\\[','',opts_json)
opts_json <- gsub('\\]','',opts_json)

# Make the options
# No data needed
input_json <- paste0('{"opts":', opts_json, '}' )

# Send the API request
results_json <- POST(url = url,
                  add_headers('Content-Type' = 'application/json'),
                  add_headers('Accept' = 'application/json'),
                  add_headers('charset' = 'UTF-8'),
                  body = input_json,
                  encode = "json"
                  )

# Convert JSON results to a data frame
results <- as.data.frame( fromJSON( rawToChar( results_json$content ) ) )

# Display the results
print( results )

#################################
# Example 9: GNRS project contributors
# (For acknowledgements)
#################################
rm( list = Filter( exists, c("results", "results_json") ) )

# Set mode
mode <- "collaborators"		

# Re-form the options json again
# Note that only 'mode' is needed
opts <- data.frame(c(mode))
names(opts) <- c("mode")
opts_json <- jsonlite::toJSON(opts)
opts_json <- gsub('\\[','',opts_json)
opts_json <- gsub('\\]','',opts_json)

# Make the options
# No data needed
input_json <- paste0('{"opts":', opts_json, '}' )

# Send the API request
results_json <- POST(url = url,
                  add_headers('Content-Type' = 'application/json'),
                  add_headers('Accept' = 'application/json'),
                  add_headers('charset' = 'UTF-8'),
                  body = input_json,
                  encode = "json"
                  )

# Convert JSON results to a data frame
results <- as.data.frame( fromJSON( rawToChar( results_json$content ) ) )

# Display the results
print( results )

